lib/pure/strutils.nim

```diff
@@ -1374,6 +1354,47 @@ proc format*(formatstr: string, a: varargs[string, `$`]): string {.noSideEffect,
   result = newStringOfCap(formatstr.len + a.len)
   addf(result, formatstr, a)
 
+# Moved down here because the macros module requires cmpIgnoreStyle and format
+import macros
+
+proc parseEnum*[T: enum](s: string): T =
+  ## Parses an enum ``T``.
+  ##
+  ## Raises ``ValueError`` for an invalid value in `s`. The comparison is
+  ## done in a style insensitive way.
+  macro m: stmt =
+    result = newNimNode(nnkCaseStmt).add(newDotExpr(newIdentNode("s"),
+      newIdentNode("normalize")))
+
+    for e in T:
+      result.add(newNimNode(nnkOfBranch).add(newStrLitNode(($e).normalize),
+        newNimNode(nnkReturnStmt).add(newIdentNode($e))))
+
+    result.add(newNimNode(nnkElse).add(newNimNode(nnkRaiseStmt).add(newCall(
+      newIdentNode("newException"), newIdentNode("ValueError"),
+      newNimNode(nnkInfix).add(newIdentNode("&"),
+        newStrLitNode("invalid enum value: "), newIdentNode("s"))))))
+
+  m()
+
+proc parseEnum*[T: enum](s: string, default: T): T =
+  ## Parses an enum ``T``.
+  ##
+  ## Uses `default` for an invalid value in `s`. The comparison is done in a
+  ## style insensitive way.
+  macro m: stmt =
+    result = newNimNode(nnkCaseStmt).add(newDotExpr(newIdentNode("s"),
```

Comment:
> I'm wondering if I could move the macros outside of the proc, but didn't succeed.

Mainly because of the `T`, getting into a macro doesn't seem that easy.